python-package/lightgbm/callback.py

```diff
@@ -209,12 +215,69 @@ def _init(env):
+                        ("map", {"mean_average_precision"}),
+                        ("binary_logloss", {"binary"}),
+                        ("multi_logloss",
+                        {"multiclass", "softmax", "multiclassova", "multiclass_ova", "ova", "ovr"}),
+                        ("xentropy", {"cross_entropy"}),
+                        ("xentlambda", {"cross_entropy_lambda"}),
+                        ("kldiv", {"kullback_leibler"})]
+        for metric, aliases in metrics_list:
+            if target_metric in aliases:
+                return metric
+        return target_metric
+
     def _callback(env):
         if not cmp_op:
             _init(env)
         if not enabled[0]:
             return
+        if first_metric_only:
+            # Choosing a target metric for early stopping from first element of metrics list as `eval_metric`
+            # if first_metric_only == True
+            eval_metric = None
+            for metric_alias in ['metric', 'metrics', 'metric_types']:
+                if metric_alias in env.params.keys():
+                    if isinstance(env.params[metric_alias], (tuple, list)):
+                        metric_list = [m for m in env.params[metric_alias] if m is not None]
+                        eval_metric = metric_list[0]
+                    else:  # string
+                        if env.evaluation_result_list[0][0] == "cv_agg":  # for lgb.cv
+                            for i in range(len(env.evaluation_result_list)):
+                                # Removing training data
+                                if ((env.evaluation_result_list[i][1].find(" ") >= 0
+                                     and env.evaluation_result_list[i][1].split(" ")[0] != "train")):
+                                    eval_metric = env.evaluation_result_list[i][1].split(" ")[1]
+                                    break
+                        else:  # the other than lgb.cv
+                            eval_metric = env.params[metric_alias]
+                    break
+            if eval_metric is None:
+                # If `eval_metric` was not chosen from `env.params`,
+                # using first element of env.evaluation_result_list for early stopping
+                eval_metric = env.evaluation_result_list[0][1]
+            eval_metric = _metric_alias_matching(eval_metric)
         for i in range_(len(env.evaluation_result_list)):
+            metric_key = env.evaluation_result_list[i][1]
+            if env.evaluation_result_list[i][0] == "cv_agg":  # for lgb.cv
+                if metric_key.split(" ")[0] == "train":
+                    continue  # train metric is not used on early stopping
+                if ((first_metric_only and eval_metric is not None and eval_metric != ""
+                     and metric_key != "valid {}".format(eval_metric) and metric_key != eval_metric)):
+                    continue
+            elif env.evaluation_result_list[i][0] == "training":
```

Comment:
I think it should be `env.model._train_data_name` (and remove one underscore from `basic.py`)
https://github.com/microsoft/LightGBM/blob/1bd15b9447e0925e6704cfce6caa3ba4fa9b7758/python-package/lightgbm/basic.py#L1600